File: test/corpus/propagation/fallible_option_some.exs
defmodule RustQ.Corpus.Propagation.FallibleOptionSome do
  @moduledoc "Propagate fallible values nested inside Some construction."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec decode_rect(term()) :: R.nif_result(R.path(:Rect))
  defrust decode_rect(_term) do
    {:ok, Rect.default()}
  end

  @spec maybe_rect(term()) :: R.nif_result(R.option(R.path(:Rect)))
  defrust maybe_rect(term) do
    {:ok, some(decode_rect(term))}
  end

  @spec use_rect_option(R.option(R.path(:Rect))) :: R.nif_result(R.unit())
  defrust use_rect_option(_rect) do
    :ok
  end

  @spec maybe_rect_from_option(R.option(term())) :: R.nif_result(R.unit())
  defrust maybe_rect_from_option(term) do
    rect =
      case term do
        {:some, term} -> some(decode_rect(term))
        :none -> none()
      end

    use_rect_option(rect)
  end
end

File: test/corpus/propagation/fallible_option_some.rs
fn decode_rect<'a>(_term: Term<'a>) -> NifResult<Rect> {
    Ok(Rect::default())
}

fn maybe_rect<'a>(term: Term<'a>) -> NifResult<Option<Rect>> {
    Ok(Some(decode_rect(term)?))
}

fn use_rect_option(_rect: Option<Rect>) -> NifResult<()> {
    Ok(())
}

fn maybe_rect_from_option<'a>(term: Option<Term<'a>>) -> NifResult<()> {
    let rect = match term {
        Some(term) => Some(decode_rect(term)?),
        None => None,
    };
    use_rect_option(rect)
}
